File: LiquidDemocracyJava/src/com/testchallenge/UserInputController.java
package com.testchallenge;

import java.io.*;
import java.util.*;

/** Decodes the use input and extracts the commands for voters */
class UserInputController {
  private final IVoterController voterController;
  private final IPrinter printer;

  /**
   * UserInputController constructor with injected voterController and printer with the interfaces
   *
   * @param voterController
   * @param printer
   */
  public UserInputController(IVoterController voterController, IPrinter printer) {
    this.voterController = voterController;
    this.printer = printer;
  }

  /** Runs the main loop that reads a user input, and calls parse and decode methods */
  public void run() {
    final String initialMessage =
        "Please, enter the command as follows: VOTER COMMAND [PICK or DELEGATE]. Press q for exit.";
    Console console = System.console();
    Scanner scanner = new Scanner(System.in);
    while (true) {
      System.out.println(initialMessage);
      String commandLine = scanner.nextLine();
      String[] splittedLine = commandLine.split(" ");
      switch (splittedLine.length) {
        case 1:
          if (splittedLine[0].toUpperCase().equals("Q")) {
            return;
          }
          break;
        case 3:
          if (decodeSplittedLine(splittedLine)) {
            printer.printItems(voterController.getAggregatedVoterChoicesForValidVotes());
            printer.printLine(voterController.getInvalidVotesNumber() + " invalid");
          }
          break;
        default:
          break;
      }
    }
  }

  /**
   * Decodes the splitted input line with calling corresponding voterController methods.
   *
   * @param splittedLine
   * @return A boolean value that indicates if the decoding process was successful.
   */
  private boolean decodeSplittedLine(String[] splittedLine) {
    String voterName = splittedLine[0];
    String commandString = splittedLine[1];
    String choiceOrDelegate = splittedLine[2];
    if (commandString.toUpperCase().equals("PICK")) {
      voterController.addVoterWithChoice(voterName, choiceOrDelegate);
      return true;
    } else if (commandString.toUpperCase().equals("DELEGATE")) {
      voterController.addVoterWithDelegate(voterName, choiceOrDelegate);
      return true;
    }
    return false;
  }
}

File: LiquidDemocracyCSharp/LiquidDemocracy/UserInputController.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace LiquidDemocracy
{
    /// <summary>
    /// Decodes the use input and extracts the commands for voters
    /// </summary>
    class UserInputController
    {
        IVoterController voterController;
        IPrinter printer;
        public UserInputController(IVoterController voterController, IPrinter printer)
        {
            this.voterController = voterController;
            this.printer = printer;
        }
        public void Run()
        {
            const String initialMessage = "Please, enter the command as follows: VOTER COMMAND [PICK or DELEGATE]. Press q for exit.";
            Console.WriteLine(initialMessage);
            while (true)
            {
                string commandLine = Console.ReadLine();
                Regex argReg = new Regex(@"\w+|""[\w\s]*""");

                int commandsCount = argReg.Matches(commandLine).Count;
                string[] splittedLine = ExtractCommandsFromCommandLine(commandLine, argReg);
                switch(splittedLine.Length)
                {
                    case 1:
                        if(splittedLine[0].ToUpper().Equals("Q"))
                        {
                            return;
                        }
                        else
                        {
                            Console.WriteLine(initialMessage);
                        }
                        break;
                    case 3:
                        if(DecodeSplittedLine(splittedLine))
                        {
                            printer.Print();
                        }
                        break;
                    default:
                        Console.WriteLine(initialMessage);
                        break;
                }
            }
        }

        string[] ExtractCommandsFromCommandLine(string commandLine, Regex argReg)
        {
            string[] commands = new string[argReg.Matches(commandLine).Count];

            int i = 0;
            foreach (var enumerator in argReg.Matches(commandLine))
            {
                commands[i++] = (string)enumerator.ToString();
            }

            return commands;
        }

        bool DecodeSplittedLine(string[] splittedLine)
        {
            String voterName = splittedLine[0];
            String commandString = splittedLine[1];
            String choiceOrDelegate = splittedLine[2];
            if (commandString.ToUpper().Equals("PICK"))
            {
                voterController.AddVoterWithChoice(voterName, choiceOrDelegate);
                return true;
            }
            else if (commandString.ToUpper().Equals("DELEGATE"))
            {
                voterController.AddVoterWithDelegate(voterName, choiceOrDelegate);
                return true;
            }
            return false;
        }
    }
}
